src/source/src/parser/pb_parser.rs

```diff
@@ -0,0 +1,305 @@
+use std::path::Path;
+
+use prost_reflect::{Cardinality, DescriptorPool, FieldDescriptor, Kind, MessageDescriptor, DynamicMessage, Value};
+use risingwave_common::error::ErrorCode::{InternalError, NotImplemented, ProtocolError};
+use risingwave_common::error::{Result, RwError};
+use risingwave_common::types::{DataType, ScalarImpl, OrderedF32, OrderedF64, Decimal};
+use risingwave_common::array::{ListValue, StructValue};
+use risingwave_pb::plan_common::ColumnDesc;
+use risingwave_expr::vector_op::cast::{str_to_date, str_to_timestamp};
+
+use url::Url;
+
+use crate::{SourceParser, SourceColumnDesc, WriteGuard, dtype_to_source_column_desc};
+
+#[derive(Debug, Clone)]
+struct ProtobufParser {
+    message_descriptor: MessageDescriptor,
+    message_name: String,
+}
+
+impl ProtobufParser {
+    pub fn new(location: &str, message_name: &str) -> Result<Self> {
+        let url = Url::parse(location)
+            .map_err(|e| InternalError(format!("failed to parse url ({}): {}", location, e)))?;
+
+        let schema_bytes = match url.scheme() {
+            "file" => {
+                let path = url.to_file_path().map_err(|_| {
+                    RwError::from(InternalError(format!("illegal path: {}", location)))
+                })?;
+
+                if path.is_dir() {
+                    // TODO(TaoWu): Allow user to specify a directory of protos.
+                    return Err(RwError::from(ProtocolError(
+                        "schema file location must not be a directory".to_string(),
+                    )));
+                }
+                Self::local_read_to_bytes(&path)
```

Comment:
I don't get it to be honest. Here we require for path to the schema file compiled by protoc

Reply:
the file may look like `SCHEMA` file in the pr, although I push it by mistake 🥵

Reply:
the content is for building a DescriptorPool and if it does not meet pb schema format, it will be discarded.

Reply:
>  But i think it's safe, we only read this file and try to parse schema from it.

Besides that, we might have DDoS attack. Users can ask us to read `/dev/stdout`, and the whole thread will stuck.

Reply:
Why don't we directly send the schema over wire protocol?

Reply:
> Why don't we directly send the schema over wire protocol?

In production environments, schema definitions are often very complex and scattered in different files, and explicitly writing out all definitions in sql is not an acceptable option.
In the current design, the user needs to compile the schema using protoc and resolve all dependencies on their own rather than on RW. We also encourage users to manage dependencies on s3 rather than locally

Reply:
Looks good overall.